tests/docs/multi_dut_asic_testing_support.md

```diff
@@ -0,0 +1,499 @@
+
+## Requirements
+
+To support multi-DUT and multi-ASIC testing, the `duthosts` and `duthost` fixture need to be updated. The target is that in test scripts we can use these fixtures to perform operations for
+* all DUTs
+* specific type of DUTs
+* specific DUT
+* all ASICs
+* specific ASIC
+* combinations of the above scenarios
+
+Meanwhile, we need to ensure backward compatibility for scenarios including:
+* single DUT, single ASIC
+* Multiple single-ASIC DUTs
+
+The test scripts should be able to use interfaces like below to perform operations on DUTs and ASICs
+```python
+duthost.foo()
+duthost.foo(asic_index=1)
+duthost.asics[1].foo()      # Equivalent to the line above
+duthost.foo(asic_inex="all")
+
+[asic.foo() for asic in duthost.asics]
+
+duthosts.foo()
+duthosts.foo(asic_index=1)
+duthosts.foo(asic_index="all")
+
+[node.foo() for node in duthosts]
+
+duthosts.nodes.foo()                 # Equivalent to: duthosts.foo()
+duthosts.nodes.foo(asic_index=1)     # Equivalent to: duthosts.foo(asic_index=1)
+duthosts.nodes.foo(asic_index="all") # Equivalent to: duthosts.foo(asic_index="all")
+
+[node.foo() for node in duthosts.nodes]
+
+duthosts[1].foo()
+duthosts[1].foo(asic_index=1)
+duthosts[1].asics[1].foo()        # Equivalent to the line above
+duthosts[1].foo(asic_index="all")
+
+duthosts["node1"].foo()
+duthosts["node1"].foo(asic_index=1)
+duthosts["node1"].asics[1].foo()        # Equivalent to the line above
+duthosts["node1"].foo(asic_index="all")
+
+duthosts.nodes[1].foo()                 # Equivalent to: duthosts[1].foo()
+duthosts.nodes[1].foo(asic_index=1)     # Equivalent to: duthosts[1].foo(asic_index=1)
+duthosts.nodes[1].asics[1].foo()        # Equivalent to: duthosts[1].asics[1].foo()
+duthosts.nodes[1].foo(asic_index="all") # Equivalent to: duthosts[1].foo(asic_index="all")
+
+duthosts.nodes["node1"].foo()                 # Equivalent to: duthosts["node1"].foo()
+duthosts.nodes["node1"].foo(asic_index=1)     # Equivalent to: duthosts["node1"].foo(asic_index=1)
+duthosts.nodes["node1"].asics[1].foo()        # Equivalent to: duthosts["node1"].asics[1].foo()
+duthosts.nodes["node1"].foo(asic_index="all") # Equivalent to: duthosts["node1"].foo(asic_index="all")
+
+duthosts.frontend_nodes.foo()
+duthosts.frontend_nodes.foo(asic_index=1)
+duthosts.frontend_nodes.foo(asic_index="all")
+
+[node.foo() for node in duthosts.frontend_nodes]
+
+duthosts.frontend_nodes[1].foo()
+duthosts.frontend_nodes[1].foo(asic_index=1)
+duthosts.frontend_nodes[1].asics[1].foo()  # Equivalent to the line above
+duthosts.frontend_nodes[1].foo(asic_index="all")
+
+duthosts.frontend_nodes["node1"].foo()
+duthosts.frontend_nodes["node1"].foo(asic_index=1)
+duthosts.frontend_nodes["node1"].asics[1].foo()  # Equivalent to the line above
+duthosts.frontend_nodes["node1"].foo(asic_index="all")
+
+duthosts.supervisor_nodes.bar()
+duthosts.supervisor_nodes[0].bar()
+duthosts.supervisor_nodes["supervisor0"].bar()
+```
+
```

Comment:
For each scenario should there be a list suggested APIs to use? If everyone uses same API it's easier to understand the code. Also becomes easier to refactor if there's such a need later.

Reply:
Thanks for the suggestion! This document proposed various ways of using APIs. For some scenarios, multiple APIs can do the same thing. It is a little bit hard to finalize such a list for each scenario before we have some real code. But sure, we'll try to improve. It would be easier if there is always single best way to do something.